libs-scala/ports/src/main/scala/com/digitalasset/ports/FreePort.scala

```diff
@@ -3,17 +3,127 @@
 
 package com.daml.ports
 
+import java.io.IOException
 import java.net.{InetAddress, ServerSocket}
 
+import com.daml.bazeltools.BazelRunfiles.rlocation
+
+import scala.io.Source
+import scala.sys.process.Process
+import scala.util.{Failure, Random, Try}
+
 object FreePort {
+  private val maxAttempts = 100
+  private[ports] val dynamicRange = dynamicPortRange()
 
+  /** Find a free port.
+    *
+    * This determines whether a port is free by attempting to bind it and immediately releasing it again.
+    * Note, this does not guarantee that the port is still free after this function returns, see [[LockedFreePort]].
+    *
+    * Draws ports from outside the dynamic port range (port 0)
+    * to avoid collisions with parts of the code base that directly bind to port 0.
+    */
   def find(): Port = {
-    val socket = new ServerSocket(0, 0, InetAddress.getLoopbackAddress)
+    val portGen = randomPortGen(dynamicRange)
+    val portCandidates = (1 to maxAttempts).map(_ => portGen())
+    portCandidates
+      .find { candidate =>
+        try {
+          val socket = new ServerSocket(candidate, 0, InetAddress.getLoopbackAddress)
+          socket.close()
+          true
+        } catch {
+          case _: IOException => false
+        }
+      }
+      .map(found => Port(found))
+      .getOrElse(throw NoFreePortException)
+  }
+
+  sealed abstract class FindPortException(message: String, cause: Throwable)
+      extends RuntimeException(message, cause) {
+    def this(message: String) = this(message, null)
+  }
```

Comment:
```suggestion
  sealed abstract class FindPortException(message: String, cause: Throwable = null)
      extends RuntimeException(message, cause)
```
This should be effectively the same.

Reply:
:+1: Was taken from somewhere else in the repo.